python/cucim/src/cucim/skimage/measure/_regionprops_gpu_misc_kernels.py

```diff
@@ -0,0 +1,721 @@
+    )
+
+    perimeter_image = ndi.convolve(
+        border_image,
+        perimeter_weights,
+        mode="constant",
+        cval=0,
+        output=cp.uint8,
+    )
+
+    min_integer_type = _get_min_integer_dtype(max_label, signed=False)
+    # if labels.dtype != min_integer_type:
+    #     labels = labels_dilated.astype(min_integer_type)
+
+    # dilate labels by 1 pixel so we can sum with values in XF to give
+    # unique histogram bins for each labeled regions (as long as no labeled
+    # regions are within < 2 pixels from another labeled region)
+    labels_dilated = ndi.grey_dilation(
+        labels, 3, mode="constant", output=min_integer_type
+    )
+
+    if robust and labels_close.size > 0:
+        if props_dict is not None and "slice" in props_dict:
+            slices = props_dict["slice"]
+        else:
+            _, slices = regionprops_bbox_coords(labels, return_slices=True)
+
+    # sum the coefficients for each label to compute the perimeter
+    perimeters = _weighted_sum_of_filtered_image(
+        label_image=labels_dilated,
+        max_label=max_label,
+        image_filtered=perimeter_image,
+        coefs=perimeter_coefs,
+        pixels_per_thread=pixels_per_thread,
+    )
+    if robust:
+        # recompute perimeter in isolation for each region that may be too
+        # close to another one
+        shape = binary_image.shape
+        for lab in labels_close:
+            sl = slices[lab - 1]
+
+            # keep boundary of 1 so object is not at 'edge' of cropped
+            # region (unless it is at a true image edge)
+            ld = labels[
+                max(sl[0].start - 1, 0) : min(sl[0].stop + 1, shape[0]),
+                max(sl[1].start - 1, 0) : min(sl[1].stop + 1, shape[1]),
+            ]
+
+            # print(f"{lab=}, {sl=}")
+            # import matplotlib.pyplot as plt
+            # plt.figure(); plt.imshow(ld.get()); plt.show()
+
+            p = regionprops_perimeter(
+                ld == lab, max_label=1, neighborhood=neighborhood, robust=False
+            )
+            perimeters[lab - 1] = p[0]
+    if props_dict is not None:
+        props_dict["perimeter"] = perimeters
+    return perimeters
+
+
+@cp.memoize(for_each_device=True)
+def _get_perimeter_crofton_weights_and_coefs(
+    directions, coefs_dtype=cp.float32
+):
+    # determine convolution weights
+    filter_weights = cp.array(
+        [[0, 0, 0], [0, 1, 4], [0, 2, 8]], dtype=cp.float32
+    )
+
+    if directions == 2:
+        coefs = [
+            0,
+            np.pi / 2,
+            0,
+            0,
+            0,
+            np.pi / 2,
+            0,
+            0,
+            np.pi / 2,
+            np.pi,
+            0,
+            0,
+            np.pi / 2,
+            np.pi,
+            0,
+            0,
+            0,
+        ]
+    else:
+        sq2 = math.sqrt(2)
+        coefs = [
+            0,
+            np.pi / 4 * (1 + 1 / sq2),
+            np.pi / (4 * sq2),
```

Comment:
I’m wondering if hardcoded floating-point values (e.g., np.pi / 4 * (1 + 1 / sq2)), which may introduce minor numerical errors, could be precomputed as high-precision constants instead.

Reply:
We could, but I don't think it will make any practical difference. Python's double precision float should be accurate enough. There will be much larger error from the discrete pixel size than numerical precision loss in these values.